### ISSUE TITLE
generate an empty text event on expanding self closing tags

### ISSUE THREAD

**Issue by pacman82** [tafia/quick-xml#93]
I'd like to suggest generating empty Text events then expanding self closing tags and empty Text events are not trimmed. How do you feel about it?

**Comment by tafia** [comment]
Sorry for the very late answer. I don't know if this is still something you're interested in.
To be honest I'm not a big fan of this feature (why not a comment or something else?). May I know the context why you'd need it?

**Comment by pacman82** [comment]
I do not need this, but I am still interested in this. To give you my motivation for suggesting this. I see enabling expanding tags as a way of saying: "I just want to treat <A/> the same way I would treat <A></A>.". Yet there is a subtle difference the events which are generated for <A></A> are: open tag, text event, close tag. The events generated for <A/> are: open tag, close tag.

You may have guessed it by now. This difference in behaviour let to a bug in my code regarding self closing tags. My original logic picked up the text event following the opening tag. This meant that for self closing tags picked I saw the whitespace characters following it instead of the empty text I expected.

I have since fixed the Bug handling the closing tag event explicitly and yielding an empty string explicitly, if no text event has occurred since the last opening tag. In a way I do check for self closing tags.

I have no strong feelings about this. Peoples expectations do work differently. I expected expanding tags to emulate <A></A> which meant an extra text event in case empty texts are not trimmed. I you feel differently just close the issue.

**Comment by andy128k** [comment]
In HTML (and SGML) self closing tags are not the same as empty tags.
<script src="..."></script> cannot be written as <script src="..."/>.

I realize that this library is not about HTML, but still. I would not make any suggestions and changes here.

**Comment by pacman82** [comment]
@andy128k I agree, these are not the same. My suggestion explicitly refers to the treatment of self closing tags, **if** `expand_empty_elements` is set to true. Setting `expand_empty_elements` to true for me, implies that I want to treat self closing tags the same way I want to treat empty tags in my domain.

**Comment by andy128k** [comment]
@pacman82 Correct me if I am wrong, but there is no TextEvent for empty tag here https://github.com/tafia/quick-xml/blob/master/tests/documents/sample_3_full.txt#L10-L12

**Comment by pacman82** [comment]
@andy128k I fear I am missing the point. As I see it the `<b>` tag in that example is not empty, but contains character data. ... I can start to see there the confusion might be coming from ... . What I refer to as "empty text events" are just regular `quick_xml::events::Event::Text` those inner `BytesText` would defer to an empty `&[u8]`. I am not arguing to introduce a new `Event` type. Is this there we are missing each other?

**Comment by andy128k** [comment]
@pacman82 I'm trying to say, that there is no text event between open and close events. those characters belong to start event.

**Comment by pacman82** [comment]
@andy128k Thanks for pointing that out. I did indeed miss that. This is however different from the behavior I have seen within my application.  I guess these `Event`s were generated using `trim_text` set to true. Which would trim empty text events. This is not the case in my application however. Until you pointed this out `trim_text` has not been on my mind. As my intention is to treat `<A></A>` and `<A/>` the same **if** `expand_empty_elements` is true. I would like to refine my proposal to only emit empty text events if `trim_text` is set to false.

**Comment by andy128k** [comment]
@pacman82 What about totally opposite? Do not emit empty text events in trim_text mode.

**Comment by pacman82** [comment]
@andy128k Is this not the case already? I have not tested my application with `trim_text(true)`, so I do not have first hand experience. I think the behavior should already be fine in that case, since both `<A></A>` and `<A/>` are the same. Here is how I see the current behavior:

| trim_text | `<A><A/>`        | `<A/>`      |
|-----------|----------------|-----------|
| true      | Start,End      | Start,End |
| false     | Start,Text,End | Start,End |

**Comment by andy128k** [comment]
@pacman82 Instead of guessing I propose to start from writing a test, to see what's really going on.

**Comment by pacman82** [comment]
@andy128k My application has/is a test for the bottom false row. This is the behavior I am interested in and want to change. I would gladly make a PR and contribute such a test upstream. However if I were to write such a test I would let it fail if an empty text event is not generated for `<A/>` and `trim_text(false)`. Since there can be difference of opinion on what the correct behavior should be, my intention is to wait for an ACK from the Author in before putting more effort into it and making a contribution.
